src/Utilities.cpp

```diff
@@ -283,125 +283,125 @@ double brentRoot(
-          (!bisected && fabs(nextPoint - currentPoint) >= fabs(previousPoint - evenOlderPoint) / 2) ||
-          (bisected && fabs(currentPoint - previousPoint) < epsilon) ||
-          (!bisected && fabs(previousPoint - evenOlderPoint) < epsilon)) {
-        nextPoint = (currentPoint + counterPoint) / 2;
-        bisected = true;
-      }
-      else {
-        bisected = false;
-      }
-
-      // Setup for next iteration
-      evenOlderPoint = previousPoint;
-      previousPoint = currentPoint;
-      previousFunc = currentFunc;
-      nextFunc = func(nextPoint);
-      if (counterFunc * nextFunc < 0) {
-        currentPoint = nextPoint;
-        currentFunc = nextFunc;
-      }
-      else {
-        counterPoint = nextPoint;
-        counterFunc = nextFunc;
-      }
-    } while (++iteration < 30 && fabs(counterPoint - currentPoint) > epsilon);
+  double epsilon)
+{
+  double counterPoint = lowerBound;
+  double currentPoint = upperBound;
+  double counterFunc = func(counterPoint);
+  double currentFunc = func(currentPoint);
+  if (counterFunc * currentFunc > 0.0) {
+    throw std::invalid_argument("Function values at the boundaries have the same sign [brentRoot].");
+  }
+  if (fabs(counterFunc) < fabs(currentFunc)) {
+    std::swap(counterPoint, currentPoint);
+    std::swap(counterFunc, currentFunc);
+  }
+
+  double previousPoint = counterPoint;
+  double previousFunc = counterFunc;
+  double evenOlderPoint = previousPoint;
+  double nextPoint;
+  double nextFunc;
+  int iteration = 0;
+  bool bisected = true;
+
+  do {
+    // Inverse quadratic interpolation
+    if (counterFunc != previousFunc && counterFunc != currentFunc && currentFunc != previousFunc) {
+      nextPoint = (counterPoint * currentFunc * previousFunc) / ((counterFunc - currentFunc) * (counterFunc - previousFunc));
+      nextPoint += (currentPoint * counterFunc * previousFunc) / ((currentFunc - counterFunc) * (currentFunc - previousFunc));
+      nextPoint += (previousPoint * currentFunc * counterFunc) / ((previousFunc - counterFunc) * (previousFunc - currentFunc));
+    }
+    // Secant method
+    else {
+      nextPoint = currentPoint - currentFunc * (currentPoint - counterPoint) / (currentFunc - counterFunc);
+    }
 
-    return nextPoint;
-  }
+    // Bisection method
+    if (((currentPoint - nextPoint) * (nextPoint - (3 * counterPoint + currentPoint) / 4) < 0) ||
+        (bisected && fabs(nextPoint - currentPoint) >= fabs(currentPoint - previousPoint) / 2) ||
+        (!bisected && fabs(nextPoint - currentPoint) >= fabs(previousPoint - evenOlderPoint) / 2) ||
+        (bisected && fabs(currentPoint - previousPoint) < epsilon) ||
+        (!bisected && fabs(previousPoint - evenOlderPoint) < epsilon)) {
+      nextPoint = (currentPoint + counterPoint) / 2;
+      bisected = true;
+    }
+    else {
+      bisected = false;
+    }
 
-double secantRoot(double lowerBound, double upperBound, std::function<double(double)> func,
```

Comment:
Is this something we'd ever want to use elsewhere, if not here?

Reply:
No, Brent's is superior in the generic function space and Newton's is superior for polynomials.

Reply:
Sounds good to me!